flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/StagewiseSchedulingStrategy.java

```diff
@@ -0,0 +1,156 @@
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * {@link SchedulingStrategy} instance which schedules tasks in granularity of stage. Note that this
+ * strategy only supports ALL_EDGES_BLOCKING batch jobs.
+ */
+public class StagewiseSchedulingStrategy implements SchedulingStrategy, SchedulingTopologyListener {
+
+    private final SchedulerOperations schedulerOperations;
+
+    private final SchedulingTopology schedulingTopology;
+
+    private final DeploymentOption deploymentOption = new DeploymentOption(false);
+
+    private final Set<SchedulingExecutionVertex> pendingVertices = new HashSet<>();
+
+    public StagewiseSchedulingStrategy(
+            final SchedulerOperations schedulerOperations,
+            final SchedulingTopology schedulingTopology) {
+
+        this.schedulerOperations = checkNotNull(schedulerOperations);
+        this.schedulingTopology = checkNotNull(schedulingTopology);
+        schedulingTopology.registerSchedulingTopologyListener(this);
+    }
+
+    @Override
+    public void startScheduling() {
+        Set<SchedulingExecutionVertex> sourceVertices =
+                IterableUtils.toStream(schedulingTopology.getVertices())
+                        .filter(vertex -> vertex.getConsumedPartitionGroups().isEmpty())
+                        .collect(Collectors.toSet());
+
+        maybeScheduleVerticesOneByOne(sourceVertices);
+    }
+
+    @Override
+    public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
+        maybeScheduleVerticesOneByOne(
+                SchedulingStrategyUtils.getVerticesFromIds(schedulingTopology, verticesToRestart));
+    }
+
+    @Override
+    public void onExecutionStateChange(
+            ExecutionVertexID executionVertexId, ExecutionState executionState) {
+        if (executionState == ExecutionState.FINISHED) {
+            SchedulingExecutionVertex executionVertex =
+                    schedulingTopology.getVertex(executionVertexId);
+
+            Set<SchedulingExecutionVertex> consumerVertices =
+                    IterableUtils.toStream(executionVertex.getProducedResults())
+                            .flatMap(partition -> IterableUtils.toStream(partition.getConsumers()))
+                            .collect(Collectors.toSet());
+
+            maybeScheduleVerticesOneByOne(consumerVertices);
+        }
+    }
+
+    @Override
+    public void onPartitionConsumable(IntermediateResultPartitionID resultPartitionId) {}
+
+    @Override
+    public void notifySchedulingTopologyUpdated(
+            SchedulingTopology schedulingTopology, List<ExecutionVertexID> newlyAddedVertices) {
+        checkState(schedulingTopology == this.schedulingTopology);
+        pendingVertices.addAll(
+                SchedulingStrategyUtils.getVerticesFromIds(
+                        schedulingTopology, new HashSet<>(newlyAddedVertices)));
```

Comment:
I think we can change `getVerticesFromIds(SchedulingTopology, Set<ExecutionVertexID>)` to be `getVerticesFromIds(SchedulingTopology, Collection<ExecutionVertexID>)` so that we do not need this conversion.